### ISSUE TITLE
External mounts without OverlayFS

### ISSUE THREAD

**Issue by zyndagj** [apptainer/singularity#1807]
## Version of Singularity:

2.5.2

### Actual behavior

When `singularity.conf` contains
```
bind path = /scratch
```
and
```
enable overlay = no
```
The external `/scratch` filesystem is available in containers with a `/scratch` directory
```
nid00031(1008)$ singularity exec good_2.5.2.img ls $SCRATCH
consulting	 ncbi  oases	   singularity_merge_mess
cyverse_scratch  ncsu  oases_test
```
and unavailable in containers without `/scratch`
```
nid00031(1008)$ singularity exec bad_2.5.2.img ls $SCRATCH
WARNING: Non existent bind point (directory) in container: '/corral-tacc'
WARNING: Non existent bind point (directory) in container: '/corral-repl'
WARNING: Non existent bind point (directory) in container: '/gpfs'
WARNING: Non existent bind point (directory) in container: '/scratch'
WARNING: Non existent bind point (directory) in container: '/work'
WARNING: Non existent bind point (directory) in container: '/home1'
ERROR  : Base home directory does not exist within the container: /home1
ABORT  : Retval = 255
```

### Suggested behavior

Always add system specified bind locations to containers so those locations are available even without overlayfs support.

### Suggested fix

Modify

- https://github.com/singularityware/singularity/blob/master/libexec/bootstrap-scripts/environment/Makefile.am

to include each system bind location. This _should_ correctly modify all images pulled from dockerhub, but would leave images from shub and pre-existing containers unhandled.

Any feedback on how to handle pre-existing and shub containers or why this would be bad?

I am happy to make these patches myself, I am just concerned about stability and security.

**Comment by jmstover** [comment]
Hi @zyndagj 

  The 'envornment' that's created is a tar file. This is done a *build* time, so there isn't a `singularity.conf` that we can expect to exist.

  `$libexecdir/singularity/bootstrap-scripts/environment.tar` is where you'll find that file that is created.

  So, you could do something like:
```
    mkdir /tmp/sing-boot
    cd /tmp/sing-boot
    tar xf /path/to/singularity/bootstrap-scripts/environment.tar
    mkdir scratch
    chown root:root scratch
       [ ... anything else in the root ... ]
    tar cvf ../environment-new.tar .
    cp -a /path/to/singularity/bootstrap-scripts/environment.tar /path/to/singularity/bootstrap-scripts/environment.tar.backup
    mv ../environment-new.tar /path/to/singularity/bootstrap-scripts/environment.tar
    cd ..
    rm -rf sing-boot/
```

  Next new build it should have the 'scratch/' directory included.

  In this case you have a $HOME under `/home1` ... without having that in the container, or having overlay support, you aren't going to be able to mount your home. And it'll error out.

  For this specific error you could try something like:

```
  mkdir /tmp/$USER-tmphome
  singularity exec -H /tmp/$USER-tmphome:/home good_2.5.2.img ls $SCRATCH
```

**Comment by zyndagj** [comment]
I configure my `singularity.conf` at build time to:

- disable overlayfs
- disable user bind locations
- automatically mount
  - /home1
  - /work
  - /scratch
  - /corral-repl
  - /corral-tacc
  - /gpfs

My center has two systems that don't support OverlayFS, so I try to make the experience consistent across each one. Modifying the `environment.tar` live is not tractable or safe on a cluster, so I would have to build a new release of the package for deployment, and I figured that

https://github.com/singularityware/singularity/blob/master/libexec/bootstrap-scripts/environment/Makefile.am

would be a good place to start. The other logical place would be in

https://github.com/singularityware/singularity/blob/development-2.x/src/lib/runtime/mounts/hostfs/hostfs.c#L158

but I don't know if that would work with read-only images.

**Comment by jmstover** [comment]
The best bet is to modify the environment.tar that gets created during build. i.e. the Makefile.am you're referencing... 
https://github.com/singularityware/singularity/blob/master/libexec/bootstrap-scripts/environment/Makefile.am

You'd patch it to add in what you need like:
```
install -d -m 0755 newroot/home1
install -d -m 0755 newroot/work
install -d -m 0755 newroot/scratch
install -d -m 0755 newroot/corral-repl
install -d -m 0755 newroot/corral-tacc
install -d -m 0755 newroot/gpfs
```

That will include those directories in the environment.tar and so when an image is built, they'll be added to it.

**Comment by zyndagj** [comment]
Correct.

Any ideas about how to handle pre-existing containers or those pulled from singularity hub and any possible security concerns?

**Comment by jmstover** [comment]
Without overlay support, nope. SHUB are downloading _images_ and not building a clean image as from docker.  Even with already existing containers you can't modify them (easily) without overlayfs. You'd run into the same consistency issue that writable ext3 images were hitting and end up with possible corruption in the images.

The only possible concern I could see would be a possible symlink attack knowing what's in the environment.tar.

*Edit:* writable ext3 images

**Comment by zyndagj** [comment]
I would customize those mounts to each specific system so they _hopefully_ could not be redirected.

It also sounds like there is no easy way to port this functionality to existing shub and ext4 images. Most of our users are taking advantage of the Docker community, so this should be good enough until all of our systems eventually support OverlayFS.

Thanks for the feedback.

**Comment by zyndagj** [comment]
Closing issue. Anyone else should re-open if they have anything to add.

**Comment by DrDaveD** [comment]
@zyndagj You could include the underlay pr #1817

**Comment by zyndagj** [comment]
@DrDaveD Can you tell me more about what undermount does? It looks like uses the internal `singularity_mount` for all mounts. Any reason why this isn't currently utilized on systems without overlayfs support?

**Comment by DrDaveD** [comment]
There's a good description of what it does in the issue #1207 that first proposed it.  Let me know if you still have questions after reading that.  It wasn't included in the official singularity-2.6.0 because the singularity core team didn't want to take away time from their 3.0 development to review it.  It is included in the EPEL & Fedora build of singularity-2.6.0 because I maintain those and I had many users asking for it, so it will get plenty of production experience soon.  The singularity core team has encouraged me to port the feature to 3.0, and I plan to do it.
